File: python/linked-lists/find-the-duplicate-number.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        fast, slow = 0, 0
        first = True
        while slow != fast or first:
            slow = nums[slow]
            fast = nums[nums[fast]]
            first = False

        fast = 0
        first = True
        while fast != slow or first:
            slow = nums[slow]
            fast = nums[fast]
            first = False

        return slow

File: c++/linked-list/find-the-duplicate-number.cc
class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        int fast = 0;
        int slow = 0;
        bool first = true;

        while (first || fast != slow) {
            first = false;
            fast = nums[nums[fast]];
            slow = nums[slow];
        }

        slow = 0;

        while (fast != slow) {
            fast = nums[fast];
            slow = nums[slow];
        }

        return fast;
    }
};
